### PR TITLE
feat: redirect

### PR DESCRIPTION
**Pull request by ronag**
Continuing on from: #600

Trying to figure out the API. Doing this through `Agent` or `Pool` doesn't quite make sense IMO.

### CODE REVIEW

**Comment by mcollina** [comment]
We are not in agreement. Using an Agent will make adding redirect being transparent for `fetch()` and other high-level libraries. Adding a different API would provide a worse developer experience (however it would be a simpler implementation).

**Comment by ronag** [comment]
Hm. I agree it would
be nicer to implement this just another client and agent. I just don’t see how to implement this in a way that makes sense in Agent or Pool. 

I have some more ideas here regarding developer experience so I’ll continue working on this and see if we can reach an agreement.

**Comment by ronag** [comment]
Agent would have to return a redirect client which would have to have a reference to the agent.

Then the client needs to correctly implement all the client APIs. Many of which would not make any sense at all. Especially with cross origin in mind.

**Comment by ronag** [comment]
@mcollina PTAL. Has a change to the Agent API to enable better composition.

**Comment by ronag** [comment]
@ShogunPanda

File: lib/redirect-agent.js
Comment (ShogunPanda) [inline lib/redirect-agent.js:89]: I wouldn't do this. This will make impossible for users to detect if there was any redirect cycle.
Let the users deal with it. Since we have maxRedirections we don't really care.

File: lib/redirect-agent.js
Comment (ShogunPanda) [inline lib/redirect-agent.js:119]: Little note here.
If you trigger dispatch from here, it means you will wait for the body (which, even not common, are legit) to be transferred before following the redirect.

That's why in #600 I dispatched from `onHeaders`. Especially when going cross-origin, we don't care about the original response body, isn't it?

File: lib/redirect-agent.js
Comment (ShogunPanda) [inline lib/redirect-agent.js:35]: Shouldn't it be `!== 0`? Or am I missing something?

**Comment by ShogunPanda** [comment]
@ronag I like the API. Nice work.
I would suggest to copy tests from my PR, adapt them (basically replacing `RedirectPool` with `RedirectAgent` and see it they passes.

File: lib/redirect-agent.js
Comment (ronag) [reply lib/redirect-agent.js:119]: Unfortunately we need to receive the entire body or pipelining won't work.

File: lib/redirect-agent.js
Comment (ShogunPanda) [inline lib/redirect-agent.js:11]: Just a confirmation question here: by doing this, you allow `maxRedirections` to be specified both at agent and single request level, isn't it?

File: lib/redirect-agent.js
Comment (ronag) [reply lib/redirect-agent.js:35]: If the stream is empty (i.e. ` === 0´) we just simply destroy and remove it. So it's correct. This way we can support `pipeline`.

File: lib/redirect-agent.js
Comment (ronag) [reply lib/redirect-agent.js:11]: Correct.

File: lib/redirect-agent.js
Comment (ronag) [reply lib/redirect-agent.js:91]: How will the user deal with it though? We would need some kind of hook in that case.

File: lib/redirect-agent.js
Comment (ShogunPanda) [reply lib/redirect-agent.js:119]: Right point. 

I guess we need two approaches:

1. Same origin -> dispatch from onComplete
2. Cross origin -> dispatch from onHeaders

This way the client will experience the best performance. Am I wrong?

**Comment by ronag** [comment]
> I would suggest to copy tests from my PR, adapt them (basically replacing RedirectPool with RedirectAgent and see it they passes.

Yes, I was hoping we could collaborate on the tests. I'll port them over from your PR once we know whether @mcollina agrees on the approach.

File: lib/redirect-agent.js
Comment (ShogunPanda) [reply lib/redirect-agent.js:91]: Well, we have `maxRedirections`, isn't? If they get in a cycle at some point it will stop following ending with a 3xx.
And then they'll check history in the callback data to deal with it. But at undici level we don't care. Or not?

File: lib/redirect-agent.js
Comment (ronag) [reply lib/redirect-agent.js:119]: Yes, I think we are both right actually. We must read the body, BUT you are right that we could actually dispatch the follow immediately in `onHeaders`.

Comment (ronag) [reply lib/redirect-agent.js:119]: The only problem with dispatching in onHeaders is that I'm unsure how to handle any error to might occur while reading the original request.

Comment (ShogunPanda) [reply lib/redirect-agent.js:119]: Yeah, I come to the same conclusion XD

**Comment by ShogunPanda** [comment]
> > I would suggest to copy tests from my PR, adapt them (basically replacing RedirectPool with RedirectAgent and see it they passes.
> 
> Yes, I was hoping we could collaborate on the tests. I'll port them over from your PR once we know whether @mcollina agrees on the approach.

Wonderful! ^^

File: lib/redirect-agent.js
Comment (ronag) [reply lib/redirect-agent.js:119]: I'm a little unsure whether the added complexity is worth the performance gain which I'm not sure whether are relevant in practice (though I have no numbers to back my scepticism).

Comment (ronag) [reply lib/redirect-agent.js:119]: Usually a redirect request will receive the EOF in the same TCP packet so `onComplete` is called immediatly.

Comment (ShogunPanda) [reply lib/redirect-agent.js:119]: To be honest I have no idea as well. 

It's unlikely that servers will send big bodies in case of 3xx. But I don't trust any server XD

File: lib/redirect-agent.js
Comment (ronag) [reply lib/redirect-agent.js:91]: That's a valid point. Where do we put the history though. There is no generic callback here. Does the http standard provide any header (such as `"via"`) where we can put the history?

@mcollina wdyt?

File: lib/redirect-agent.js
Comment (ShogunPanda) [reply lib/redirect-agent.js:35]: So in that case I would do a noop in `data` and ignore the `error`, as we don't probably care as well.

Comment (ronag) [reply lib/redirect-agent.js:35]: Yea, neither of them should be ever invoke. I'm mostly guarding against broken streams in the callbacks.

File: lib/redirect-agent.js
Comment (ronag) [reply lib/redirect-agent.js:119]: You ok with leaving as is?

Comment (ShogunPanda) [reply lib/redirect-agent.js:119]: If body is empty yes.

But I do also see you're point about error handling.

Well, then maybe just leave a comment note that this can technically be done so we can revisit in the future.

File: lib/agent.js
Comment (mcollina) [inline lib/agent.js:35]: Can we keep the `get`? It will be used for #587

**Review by mcollina — COMMENTED** [review]
The approach looks good to me, I've left a comment.

File: lib/redirect-agent.js
Comment (ShogunPanda) [reply lib/redirect-agent.js:119]: Sorry, I replied before reading your last question.

Yes, I'm fine with leaving like it. Maybe, as I said above, leave a note about this possible cross-origin optimization.

File: lib/redirect-agent.js
Comment (ShogunPanda) [reply lib/redirect-agent.js:91]: As you can see in my PR, I expanded `onHeaders` in all resources so that can accept additional arguments. And I used to pass the history as it builds the callback data for all methods.

This probably is also a good idea for future expandability as new `Agent`s (either in undici, or on user-level) might benefit for it to attach custom callback data.

File: lib/redirect-agent.js
Comment (ShogunPanda) [reply lib/redirect-agent.js:35]: You're right. Better not to ignore broken stuff in the wild because they might cause leaks in users' software.

Comment (ShogunPanda) [reply lib/redirect-agent.js:35]: In that case check what happens when the error is thrown. I think it should propagate up to the user level.

File: lib/redirect-agent.js
Comment (ShogunPanda) [reply lib/redirect-agent.js:11]: Much better than mine then! ^^

File: lib/agent.js
Comment (ronag) [reply lib/agent.js:35]: Not really... the api becomes weird. I can try to give it more thought.

File: lib/redirect-agent.js
Comment (ronag) [reply lib/redirect-agent.js:35]: I'm not sure we can propagate it to the user. In this case the best we can do is unhandled exception. Since it should never happen I think that's maybe fine. @mcollina wdyt?

File: lib/agent.js
Comment (ronag) [reply lib/agent.js:35]: @mcollina I don't see how we can keep this... it just won't work with `RedirectAgent` and gives more api area that needs to be fulfilled...

File: lib/redirect-agent.js
Comment (ronag) [reply lib/redirect-agent.js:35]: I can maybe fix this.

**Comment by ronag** [comment]
I think this is as much as I can do at the moment. @ShogunPanda do you think you could help out with test coverage and docs? @Ethan-Arrowood maybe you could also help with docs?

Basically what we know how architecturewise is:

```
class Dispatch implements EventEmitter {
  dispatch (opts, handler)
  close ()
  destroy ()
}

class Agent implements Dispatch {
}

class RedirectAgent implements Dispatch {
}

interface Client implement Dispatch {
  ... client api
}

class Client implements Client {
}

class Pool implements Client {
}
```

File: lib/agent.js
Comment (mcollina) [reply lib/agent.js:35]: This was not a problem with the other implementation. Moreover this remove observability, as there is no way to know to access the Pool object.

I really think this should be retained.

Comment (ronag) [reply lib/agent.js:35]: > This was not a problem with the other implementation

The other implementation is not correct IMHO.

Comment (ronag) [reply lib/agent.js:35]: > Moreover this remove observability, as there is no way to know to access the Pool object.

I'm not sure I follow. What does this remove? Why would you need to access the Pool object?

Comment (ronag) [reply lib/agent.js:35]: It makes sense to me to leave the dispatching of a request as an implementation detail of the agent.

Comment (ronag) [reply lib/agent.js:35]: If we drop cross origin support it could make sense to keep the existing API. But that seems like a rather large thing to not support and kind of negates the point of having it inside of undici if people have to reimplement it anyway with cross origin support.

Comment (mcollina) [reply lib/agent.js:35]: I don't really understand what's the problem in exposing a get method.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: lib/agent.js
Comment (ronag) [reply lib/agent.js:35]: What would `RedirectAgent.get` return?

File: lib/agent.js
Comment (ronag) [reply lib/agent.js:66]: They are passed. Se the constructor. Factory is a closure.

```
this[kFactory] = (origin) => factory(origin, opts)
```

File: lib/agent.js
Comment (mcollina) [reply lib/agent.js:35]: `this[kAgent].get(origin)` would do.

**Review by dnlup — COMMENTED** [review]
Assuming we don't have to follow redirection using a specific origin, hence `Pool` (which I think it's unlikely, but I am just thinking out loud), I like this approach. I would expose the `get` method though, at this stage we need it and I think it causes no harm. I am open to revisiting it in the future (as I understand better what high-level and low-level APIs are needed by the user).

File: lib/agent.js
Comment (mcollina) [reply lib/agent.js:35]: Note that exposing `get(origin)` enable somebody to access some stats about the `Pool` and monitor its state.

**Comment by ShogunPanda** [comment]
@dnlup I think we are following cross-origin. I'll confirm when I import tests from #600 

@ronag I can help with test at the moment. One thing I noticed is that you completed remove the support for the history of redirections. May I ask you why?

**Comment by ronag** [comment]
> I can help with test at the moment. How can add test to your branch? Fork under my name and then send a PR for it?

Exactly!

> One thing I noticed is that you completed remove the support for the history of redirections. May I ask you why?

I was unsure about the API for it. Do you think we could add it in a follow up PR? I think I would like to discuss and iterate over that a little more.

File: lib/agent.js
Comment (ronag) [reply lib/agent.js:35]: Hm. I guess we could do that. One important note though is that `RediretPool.get(origin).request(opts)` won't follow any redirects. Which I think is kind of weird/confusing.

@dnlup also

**Comment by ShogunPanda** [comment]
> Do you think we could add it in a follow up PR?

Yes, definitely. Once we get this merge I'd like to directly switch to it once we agree on the API. 

While is not required to provide the redirect functionality, I think is really important for developer experience.

File: lib/agent.js
Comment (ronag) [reply lib/agent.js:35]: Adding @dnlup comment here so we have the whole conversation:

> Assuming we don't have to follow redirection using a specific origin, hence Pool (which I think it's unlikely, but I am just thinking out loud), I like this approach. I would expose the get method though, at this stage we need it and I think it causes no harm. I am open to revisiting it in the future (as I understand better what high-level and low-level APIs are needed by the user).

**Comment by dnlup** [comment]
> @dnlup I think we are following cross-origin. I'll confirm when I import tests from #600

Yes, we are. I explained myself poorly. I meant to say that implementing this using the `Agent` is cleaner than implementing it in the `Pool`. Unless there is a specific need to have a 'low-level `Pool` instance that can follow redirection (with the caveat that are only on the same origin). For example:

```js
const pool = new RedirectPool()

pool.request(...)
```

I think we can discard this use case because if a user needs to follow redirection, they would use the `request` and other helpers directly exposed by `undici`.

File: lib/agent.js
Comment (ronag) [reply lib/agent.js:35]: I'm adding it back as @mcollina suggested. I'm not sure I think it's a good idea but it should not block this. I can open a separate PR where we can continue this discussion later.

Comment (mcollina) [reply lib/agent.js:35]: > Hm. I guess we could do that. One important note though is that RediretPool.get(origin).request(opts) won't follow any redirects. Which I think is kind of weird/confusing.

That would be totally ok for me as long as we document it.

Comment (dnlup) [reply lib/agent.js:35]: I understand your point, @ronag , and you're right. I wouldn't use a `Pool` instance for performing requests that need redirections. I agree that it wouldn't make sense. But `get(origin)` might get used for other low-level operations that do not require a request to be fired. As of now, I see more harm than good in removing it for that reason.

> Which I think is kind of weird/confusing.

On one side, it is. On the other one, if we think in therm of low-level API, it's just a small piece that does exactly one thing, and you might need it at that level. The end-user would use the high-level API and should not be concerned about it. I think we can document it well enough for someone who wants to implement a custom `Agent`. I am sure we can come up with a better approach in the future if needed.

**Comment by ronag** [comment]
@dnlup I had to comment out a few of your `Agent` tests that failed and I don't understand how they should/could work. Do you want to take a look or should I remove them?

File: lib/agent.js
Comment (ronag) [reply lib/agent.js:35]: I know some of my colleagues that use `Agent.get` directly instead of the helper methods (they want to use dispatch). I could easily see that they would get confused and start using `RedirectAgent.get` thinking it does something it does not. Most people are also not very meticulous with reading the docs.

That's my main concern. It's kind of easy do things wrong which for me is a sign of a bad API. But anyways, let's discuss this in a separate PR. It's not that big of a problem.

**Comment by dnlup** [comment]
> @dnlup I had to comment out a few of your `Agent` tests that failed and I don't understand how they should/could work. Do you want to take a look or should I remove them?

I'll take a look at it, I think you are talking about the tests for `connect/disconnect` events right?

File: lib/agent-redirect.js
Comment (ShogunPanda) [inline lib/agent-redirect.js:77]: Little nit here. If the original method was `HEAD`, it must not be changed to get. So the line before should check for `statusCode === 303 && this.opts.method !== 'GET'`

File: docs/api/Agent.md
Comment (ShogunPanda) [inline docs/api/Agent.md:14]: Shouldn't this be `Agent.get(origin)`?

File: test/agent.js
Comment (dnlup) [inline test/agent.js:161]: @ronag In this test what fails is the check after calling `Agent.get(origin)`. The previous implementation attached the event listeners when calling `get`:

https://github.com/nodejs/undici/blob/main/lib/agent.js#L52

Comment (ronag) [reply test/agent.js:161]: Not sure I follow... nothing has changed there in this PR?

File: lib/redirect-agent.js
Comment (ronag) [reply lib/redirect-agent.js]: We agreed we would do this in a follow up PR.

**Comment by ronag** [comment]
@Ethan-Arrowood this changes the architecture a bit and and some new nice interfaces/traits. When you have time could we have a call and discuss? I think I need your help to sort out typescript and docs accordingly.

**Comment by ronag** [comment]
Question here. When the redirect cannot be followed (i.e. due to request body or too many redirects). Should we error or just continue dispatch without redirect?

**Comment by ronag** [comment]
@ShogunPanda Do we need anything else here? https://github.com/follow-redirects/follow-redirects/blob/main/index.js 

e.g. https://github.com/follow-redirects/follow-redirects/blob/main/index.js#L370-L372

**Comment by ShogunPanda** [comment]
No, it seems we implemented everything. Dropping `Authorization` header is not in the spec and I'm not sure why they did it.
I already (and thus you) implemented that 303 logic.

About your question above: IMHO once we exhaust maxRedirections, just continue as redirect did not exist and return the last 3xx response to the client with no errors.

File: lib/agent.js
Comment (mcollina) [inline lib/agent.js:50]: this should return the same instance that is being cached.

**Review by mcollina — CHANGES_REQUESTED** [review]

**Comment by ronag** [comment]
> I already (and thus you) implemented that 303 logic.

Ah, I have to make sure to add you as author to commits.

File: lib/agent.js
Comment (ronag) [reply lib/agent.js]: Good catch!

File: test/agent.js
Comment (ronag) [reply test/agent.js]: Sorted. Thanks!

**Review by mcollina — APPROVED** [review]
lgtm, good work

**Comment by ShogunPanda** [comment]
> > I already (and thus you) implemented that 303 logic.
> 
> Ah, I have to make sure to add you as author to commits.

Lol. Thanks! :)

File: lib/agent-redirect.js
Comment (Ethan-Arrowood) [inline lib/agent-redirect.js:52]: In node core I've seen this implemented using a Set and a `has` look up. It might be trivially faster than an array scan; however, this list only has 6 items so its probably a negligible perf difference (wouldn't be surprised if V8 optimizes this already since the list is constant)

File: lib/agent-redirect.js
Comment (Ethan-Arrowood) [inline lib/agent-redirect.js:112]: Do we need to do something like:

```suggestion
    this[kAgent].on('connect', (...args) => this.emit('connect', ...args))
```

File: lib/client-pool.js
Comment (Ethan-Arrowood) [inline lib/client-pool.js:158]:
```suggestion
        throw new InvalidArgumentError('origin')
```
spelling maybe

**Comment by ronag** [comment]
I found some problems:

- close/destroy will do so on the passed agent (which could also be the globalAgent).

Maybe better to just inherit `Agent` and override `dispatch`?

File: lib/core/client.js
Comment (Ethan-Arrowood) [inline lib/core/client.js:247]:
```suggestion
        throw new InvalidArgumentError('origin')
```

spelling

**Review by Ethan-Arrowood — CHANGES_REQUESTED** [review]

File: lib/agent-redirect.js
Comment (ronag) [reply lib/agent-redirect.js:112]: True.

File: docs/api/Agent.md
Comment (ronag) [reply docs/api/Agent.md:8]: Not sure. Can you make a PR suggestion?

**Comment by ronag** [comment]
> I found a possible proble. close/destroy will do so on the passed agent (which could also be the globalAgent).
> 
> I'm a little unsure how to resolve this? @mcollina @dnlup @ShogunPanda.
> 
> Alternatives:
> 
> * Don't allow passing custom agent and always use global or a new Agent instance.
> * Remove close/destroy from agent API and make it instance specific?
> * Make close/destroy noops?

Resolved by reduce required Agent interface.

File: lib/agent-redirect.js
Comment (ronag) [inline lib/agent-redirect.js:115]: What if globalAgent is also a redirect agent or self?

File: lib/agent-redirect.js
Comment (ronag) [inline lib/agent-redirect.js:122]: @mcollina wdyt?

Comment (ShogunPanda) [reply lib/agent-redirect.js:122]: Can you clarify what is this for?

Comment (ronag) [reply lib/agent-redirect.js:122]: It is possible to create a loop if you pass a redirect agent into a redirect agent.

Comment (ShogunPanda) [reply lib/agent-redirect.js:122]: Oh, I see. Cool.

**Comment by ShogunPanda** [comment]
@ronag I'm writing tests but I'm bit confused on how to use RedirectAgent with the new API.
Do you mind sharing a little (like 4 lines) snippet of an example of how to use it?

**Comment by ronag** [comment]
```js
const agent = new RedirectAgent()
undici.request(url, { ...opts, agent })
```

**Comment by ShogunPanda** [comment]
> ```js
> const agent = new RedirectAgent()
> undici.request(url, { ...opts, agent })
> ```

That's what I thought and tried. But the test failed on the circular dispatch assert.

**Comment by ronag** [comment]
maybe a bug, i'll try fix

**Comment by ronag** [comment]
can you try removing the assert?

**Comment by ShogunPanda** [comment]
> can you try removing the assert?

Yes, that made the trick. Don't we need the assert at all? To be honest I don't full understand how the loop is created if you pass a redirect agent to a redirect agent. 

Are you talking about passing `RedirectAgent 1` to `RedirectAgent 2`? In my case `RedirectAgent 1` was passed to `RedirectAgent 1`. Which made the assert fail. Now, since we have maxRedirections, why do we care about loops at all?

**Comment by ronag** [comment]
> Which made the assert fail. Now, since we have maxRedirections, why do we care about loops at all?

This is a clear user error where I think we should be helpful and fail.

**Comment by ShogunPanda** [comment]
I see. So the assert is indeed needed.
And I also think I found the problem:

```
Symbol('tag') == Symbol('tag')
> false
```

**Comment by ronag** [comment]
> I see. So the assert is indeed needed.
> And I also think I found the problem:
> 
> ```
> Symbol('tag') == Symbol('tag')
> > false
> ```

Not sure that is what is happening though.

```js
const tag =  Symbol('tag')
tag === tag
```

**Comment by ShogunPanda** [comment]
Nevermind. It's not about symbols. 

It's about detecting that the tag is not set (first redirect or no RedirectAgent) or set to the same value as the previous one (which means recursion, not loop):

```
assert(!opts[kRedirectTag] || (opts[kRedirectTag] === this[kRedirectTag]), 'circular dispatch')
```

What do you think?

File: lib/agent-redirect.js
Comment (ronag) [inline lib/agent-redirect.js:121]:
```suggestion
  assert(!opts[kRedirectTag] || (opts[kRedirectTag] === this[kRedirectTag]), 'circular dispatch')
```

**Comment by ShogunPanda** [comment]
@ronag 

I've added tests and send a PR to this branch: https://github.com/nodejs/undici/pull/614
All my existing tests cases were passing with no issues at all. Very good job!

The idea of skipping redirects when using Readable is very good. We just need to clearly document it.

Once that PR and this are merged, we have to discuss about how to add redirect history to callback data.

File: lib/agent.js
Comment (mcollina) [inline lib/agent.js:66]: I would just check `if (typeof opts !== 'object')`

File: test/utils/stream.js
Comment (mcollina) [inline test/utils/stream.js:1]: use strict is missing

Comment (ronag) [reply test/utils/stream.js:1]: @mcollina can we add linting for this?

**Comment by ronag** [comment]
Should we merge `RedirectAgent` into `Agent`? The functionality is kind of trivial and it would have to be explicitly enabled by passing the `maxRedirects` option in the constructor.

@mcollina @dnlup @ShogunPanda  wdyt?

**Comment by ShogunPanda** [comment]
I agree. 
Since it's a very common use case when dealing with HTTP agents, it's better to keep things simple for developer experience (in this case, not having to remember to use a different agent for a comon case).
And since it's protected by maxRedirects I don't see performance problems.

Also, I have to agree since it was similar to my original proposal in https://github.com/nodejs/undici/issues/553#issuecomment-781994003. Just joking. Lol.

**Comment by mcollina** [comment]
Good for me!

**Comment by ronag** [comment]
Done!

**Comment by ronag** [comment]
This is starting to feel pretty solid. Just waiting for doc + coverage. @Ethan-Arrowood @ShogunPanda.

File: lib/agent/redirect.js
Comment (ronag) [inline lib/agent/redirect.js:87]: @ShogunPanda do we have a spec ref in regards to what to do with body?

Comment (ShogunPanda) [reply lib/agent/redirect.js:87]: So, I've read the spec again. RFC 7231, Section 6.4 
It says that redirection is used to serve the requested resource from another URL, I think it assumes that no body is generated (and thus can be ignored). Even thought generating a body is not prohibited. There is an exception to this, the status 300.

For status 301, 302, 303, 307 and 308 (the latter from RFC 7238), the spec(s) mentioned that the body usually (which means it's optional and not mandated) contain just an hyperlink to the value of the Location response header, so probably we can just ignore it.

For status 300, which is "Multiple Choices", the spec mentions both generating a Location response header AND a response body with the other possible location to follow. The interesting thing is that the spec explicitily chooses not to specify a format for such body and leave it to servers and browsers implementors. Which means, in our case, that probably we can just ignore it. (we could, in theory, let the user specify a closure and let them deal with parsing, but I think it's overly complicated).

Comment (ronag) [reply lib/agent/redirect.js:87]: Can we add some notes in the code on what we do and why?

Comment (ShogunPanda) [reply lib/agent/redirect.js:87]: Done in #621

**Comment by ShogunPanda** [comment]
@ronag @mcollina 

I've sent #621 which brings coverage back to 100%. 
It also fixes couple of minor stuffs. Please let me know what you think about it.

**Comment by ShogunPanda** [comment]
Also, final note. Once this get merged, I'll start working on #615.

**Pull request merged**
